server/authz/rbac_test.go

```diff
@@ -0,0 +1,298 @@
+	// The default policy allows only GET requests
+	// So, the POST request should not be authorized
+	postRequest := mockHttpRequest("POST", "")
+	isAuth = authorizer.Authorize(postRequest, userInfo)
+	assert.False(t, isAuth)
+}
+
+// TestScopes is a test implementation of the Scopes functionality.
+// It tests that the scopes are set correctly when the authorizer is created.
+// Additionally, it tests that the required scopes are tested for the user.
+func TestScopes(t *testing.T) {
+	authorizer, err := NewCasbinObject(WithPolicyMap(testPolicyMapPath), WithPropertyFile(testPropertyFilePath))
+	assert.NoError(t, err)
+
+	// Test that the authorizer is not nil
+	assert.NotNil(t, authorizer)
+
+	// Check the scopes loaded from the policy map
+	scopes := authorizer.getCurrentScopes()
+	assert.Equal(t, 1, len(scopes))
+	assert.Equal(t, "groups", scopes[0])
+
+	userInfo := mockUserInfo(testEmail, groupReadOnly)
+
+	// get a test gin context for GET request
+	// The default policy allows only GET requests
+	// So, the GET request should be authorized
+	getRequest := mockHttpRequest("GET", "")
+	isAuth := authorizer.Authorize(getRequest, userInfo)
+	assert.True(t, isAuth)
+
+	// Change the scopes to "email"
+	authorizer.currentScopes = []string{"email"}
+	scopes = authorizer.getCurrentScopes()
+	assert.Equal(t, 1, len(scopes))
+	assert.Equal(t, "email", scopes[0])
+
+	// GET requests are for the test user as part of the read only group
+	isAuth = authorizer.Authorize(getRequest, userInfo)
+	assert.True(t, isAuth)
+
+	// POST requests are allowed for the test user email
+	postRequest := mockHttpRequest("POST", "")
+	isAuth = authorizer.Authorize(postRequest, userInfo)
+	assert.True(t, isAuth)
+
+	// PATCH requests are allowed for the test user email
+	patchRequest := mockHttpRequest("PATCH", "")
+	isAuth = authorizer.Authorize(patchRequest, userInfo)
+	assert.False(t, isAuth)
+}
+
+// TestNamespaces is a test implementation of the Namespaces based access.
+// It tests that a user can access a namespace that is in the policy map.
+func TestNamespaces(t *testing.T) {
+	authorizer, err := NewCasbinObject(WithPolicyMap(testPolicyMapPath), WithPropertyFile(testPropertyFilePath))
+	assert.NoError(t, err)
+
+	// Test that the authorizer is not nil
+	assert.NotNil(t, authorizer)
+
+	userInfo := mockUserInfo(testEmail, groupNamespace)
+
+	// Request for a namespace that is not authorized for the user
+	getRequest := mockHttpRequest("GET", "")
+	isAuth := authorizer.Authorize(getRequest, userInfo)
+	assert.False(t, isAuth)
+
+	// Request for a namespace that is authorized for the user
+	getRequest = mockHttpRequest("GET", "test_ns")
+	isAuth = authorizer.Authorize(getRequest, userInfo)
+	assert.True(t, isAuth)
+
+}
+
+// TestConfigFileReload is a test implementation of the ConfigFileReload functionality.
+// It tests that the RBAC properties are reloaded correctly when the config file is changed.
+func TestConfigFileReload(t *testing.T) {
```

Comment:
A better approach could be to verify the `Authorize` behaviour change after the config file reload instead of checking scopes.

Reply:
So once the config file is reloaded it will update the scopes and the authorize flow would be the same as the TestScopes where we check auth based on scope change, so I thought that we are testing that through the TestScopes test case. 
What do you think?

Reply:
I see. The real scenario would be there is a file reload which changes the scope and hence changes the authorization result. Hence I think a single unit test which reloads file and verifies the authorization result change will cover both file reload and scope change.

Reply:
My main point across most of the comments is to hide implementation details from unit tests. And I think things like getScope and getPolicy are all implementation details. We should test against Authorize and Authorize only.

Reply:
I am also okay with the current structure. It already covers all the cases. Feel free to disagree and merge.

Reply:
Thanks for this Keran!

I was considering the path where we would try to test the smallest units/functions using individual UTs to see if they are working as expected. 
Considering the testing of Authorize only would probably at an interface level where we just want to verify if the Auth works as expected. So maybe at an E2E level would make more sense or some UTs  at a package where we do not care about which type of Authorization implementation is used. (Like today we are just using 1 kind, we might add more impls later)
Here we might want to verify if our overall RBAC implementation is covered, like in the file reload we could have scenarios where the property is given an un-parsable value, so our code should be covering and tested for such cases using UTs.

This was my understanding for this, would love to discuss further to understand more and probably even come up with a more streamlined approach for our UTs as we go forward in our development. Given that we're 1.0 already, we definitely need to make our testing framework more robust

Reply:
Sounds good Sidhant! We can discuss this more offline. Let's make sure we have good UT coverage and at the same time, make the UTs simple.